Query Experience Cleanup Tasks

martavis opened the pull request:
Tasks completed:
- Removed a single host from target
- [Handled user login redirects better](https://github.com/fleetdm/fleet/pull/1591#pullrequestreview-728851824)
- [Added typing for Fleet Ace](https://github.com/fleetdm/fleet/pull/1591#discussion_r689742736)
- [Used `console.error` where appropriate](https://github.com/fleetdm/fleet/pull/1591#discussion_r689749642)
- [Select Targets - Round online target percentage](https://github.com/fleetdm/fleet/pull/1732#pullrequestreview-735427394)
- [Select Targets - Used given "All Linux" label from the API](https://github.com/fleetdm/fleet/pull/1732#discussion_r693257399)
- Rendered the live query warning banner when needed
- Rendered target dropdown error on error
- Rendered error when going to target selector page
- Collapsed the sidebar on page 1
- Fixed any design flaws (missing icons, spacing, etc.)
- Permissions
- Removed redux for query flow (`queryIdForEdit`, `selectedTargets`, `selectedOsqueryTable`, `currentUser`, `isBasicTier`)
- Replaced `any` types as much as possible
- Pieced together app (routes, file imports, etc)
- Removed all files no longer used pertaining to the live query flow
- Fixed tests

**This is the 4th PR for #1497, and second to last before pushing to main. The checklist will be done on the final PR.**

martavis commented:
@gillespi314 @RachelElysia @chiiph @noahtalerman 

Hey team, I need some serious support testing this overhaul. Although there are technically only 3 steps to creating and running a query, many files have been rewritten, updated, and removed; from the highest level (e.g. routes, App, AuthenticatedRoutes) to the lowest (e.g. Button). My hope is for us is the following:

- Test the save flow
- Test the run query flow
- Test the edit flow
- Use all types of users
- Use teams
- Detect any bugs
- Regression test the app in general, making sure nothing is broken
- Run through the code and give feedback

Also, FYI: I am still working on the styles since functionality was the priority, so spacing may be noticeably off.

====

As far as the code, new patterns have been used:

First, the context API is in full effect. The query page no longer uses Redux to retrieve props. Instead, you will see reference to `useContext` scattered, which gives the data necessary to a component.

Second, the Fleet API is continues to skip the Redux flow (introduced in `ManageHostsPage`). However, a newer tool used is `react-query` which allows us to fetch, cache, and sync API data much easier. You will see the use of `useQuery` scattered in the code.

In general, there is A LOT more TS typing added. I would like to formally introduce a rule for limit our types as noted below:
- Component props (interfaces)
- States
- Refs
- Method signatures params; return types at dev's discretion
- API responses

Let's discuss to get everyone's thoughts after we release 4.3.0.

Thanks everyone!

chiiph commented:
@martavis hey, not sure what happened, but the diff seems to have commits from other things.

martavis commented:
Yes, I pulled from main multiple times to stay up to date. It came in handy
too, because there were a good amount of conflicts along the way.

On Thu, Sep 2, 2021, 6:21 AM Tomas Touceda ***@***.***> wrote:

> @martavis <https://github.com/martavis> hey, not sure what happened, but
> the diff seems to have commits from other things.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fleetdm/fleet/pull/1807#issuecomment-911676000>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALG7XCOJR4IZLJ27RUX5PX3T75235ANCNFSM5C2BXQRQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

RachelElysia commented:
Ahh-- that's what I use git rebase for so it puts your commits on top of the main branch

https://git-scm.com/docs/git-rebase

I think @mikermcneil had tips on how not to commit other's commits onto your PR.

martavis commented:
Oh my, I did not see that file change count! Let me fix this up.

martavis commented:
@chiiph @RachelElysia The commits are fixed. Thanks for pointing that out.

RachelElysia commented:
Love all the deleted code! This'll definitely take me a few days to digest everything you did.

mikermcneil commented:
> I think @mikermcneil had tips on how not to commit other's commits onto your PR.

My solution is just to not use rebase 🤷

chiiph commented:
@martavis played a bit with it, overall looks great. I think it'll be an awesome improvement.

Here are the issues I saw:

- Only 1 host, I select All Hosts and it shows "1% online":

<img width="1052" alt="Screen Shot 2021-09-02 at 17 03 49" src="https://user-images.githubusercontent.com/167183/131909425-0e6f2528-f2c4-4745-9f51-16cf1270dcf2.png">

Also, if I select All hosts, it lets me select macOS and others, which feels odd.

On the oooother hand, should we not let people select labels with zero hosts?

- There doesn't seem to be a loading state, so when you run the query it looks like nothing is happening. If you wait, though, results appear. If you then click "Run again", you see this state while it's running:

<img width="1106" alt="Screen Shot 2021-09-02 at 17 04 19" src="https://user-images.githubusercontent.com/167183/131909548-c270c28d-1162-4baf-a29a-d0c0bda662cd.png">

- When you try to save without a name, it errors out as expected. But when you fill the name in, the error state remains:

<img width="861" alt="Screen Shot 2021-09-02 at 17 06 37" src="https://user-images.githubusercontent.com/167183/131909605-aa290e59-cec2-4170-b43e-41433c8c59e9.png">

martavis commented:
@noahtalerman A few product comments/questions posted above:

> Also, if I select All hosts, it lets me select macOS and others, which feels odd.

> On the oooother hand, should we not let people select labels with zero hosts?

martavis commented:
I address the issues found above minus those related to product decisions.

noahtalerman commented:
> Also, if I select All hosts, it lets me select macOS and others, which feels odd.

> On the oooother hand, should we not let people select labels with zero hosts?

@martavis let's allow the behavior in both of these comments above for now. We can improve this experience in a later iteration.

chiiph commented:
All right! This is looking better and better. Here are some other things I found:

- When searching for a host, if you click the host, then suddenly the search shows empty and it's confusing because it looks like the host disappeared, when in reality it's selected, but since it's selected you can't select it again. So I think we should dismiss the search results after you click a host.
- Once I run a query there's no way to save it?
- Once I save a query, the query appears empty right after:

<img width="614" alt="Screen Shot 2021-09-03 at 17 57 11" src="https://user-images.githubusercontent.com/167183/132064644-20c873eb-7d60-4b1d-ad52-48ebb4c8d484.png">

If I update the query and save, it works.

martavis commented:
> When searching for a host, if you click the host, then suddenly the search shows empty and it's confusing because it looks like the host disappeared, when in reality it's selected, but since it's selected you can't select it again. So I think we should dismiss the search results after you click a host.

@noahtalerman how do you feel about this? Essentially, this would mean removing the dropdown after every click. I'm not sure that's the best experience either but I'm not sure what's better.

And for context, @chiiph is referring to searching for a hostname until there is only one in the dropdown.

martavis commented:
> Once I run a query there's no way to save it?

@noahtalerman This is also a UX question.

gillespi314 reviewed (commented):
@martavis, potential bug/UX issue when creating a new query. If I attempted to delete the delete the default query, it pops right back up again in the editor and I have to delete it all over again (and again and again). If I try to select all the text and paste a query over it, the default query gets tacked back on to the end of the new query. Seems like maybe it could use a debounce?

martavis commented:
@gillespi314 Pull again for the fix.

noahtalerman commented:
> When searching for a host, if you click the host, then suddenly the search shows empty and it's confusing because it looks like the host disappeared, when in reality it's selected

Let's improve this experience in a later iteration (following the release of Fleet 4.3.0). I agree that this isn't the ideal experience.

> Once I run a query there's no way to save it?
There should be.

It looks like the current "Done" button the **Query results** page navigates the user back to the **Manage queries** page. 
@martavis, instead, let's make this button navigate the user to the **Create query** page (or **Edit query** page if the user is running a saved query) with the chosen SQL already present. The user will be able to save the query on this page.

A screenshot of the updated dev note in Figma is included below:
<img width="309" alt="Screen Shot 2021-09-07 at 3 36 27 PM" src="https://user-images.githubusercontent.com/47070608/132401028-8d38d9e1-40a5-4919-880d-c1024f7e7672.png">

martavis commented:
### TODOs based on QA

- [ ] **Targets** - Vertically center target label count
- [ ] **Query Results** - "Done" button should redirect to Create or Edit with query injected in the editor
- [ ] **Targets** - Include UUID in the tables
- [ ] Host Details - Query button should navigate to the query flow with host attached for the second step

noahtalerman commented:
@martavis the query experience is functioning, looking, and feeling **really** slick. I have three remaining requests prior to merging these changes in:
- Currently, anytime I edit an existing query's name or description on the **Edit query** page, the contents of the SQL fields are replaced with `SELECT * FROM osquery_info`. Instead, the query's SQL should remain intact when editing the query's name or description. I've recorded a Loom video to showcase this behavior:
https://www.loom.com/share/0ddb337a27d149bf96d821ebdea299ef
- Can we please add back the right side bar with the osquery documentation on the **New query** and **Edit query** pages? 
![Screen Shot 2021-09-08 at 7 13 53 AM](https://user-images.githubusercontent.com/47070608/132499703-231b5838-8af7-480b-bd2f-826f70d381dd.png)
Figma includes UI to show / hide this right sidebar. Adding this behavior can wait until the next release cycle.
- Let's only display the red bubble with the errors count if there are more than 0 errors. In the screenshot below the red bubble would be removed:
![Screen Shot 2021-09-08 at 7 18 30 AM](https://user-images.githubusercontent.com/47070608/132500268-0f651fc4-c464-4681-8e3e-9d447e2d4d51.png)
- Let's use sentence casing and display "Select targets" instead of "Select Targets" for the header on the **Select targets** page

noahtalerman commented:
I'm attempting to select targets as a global user with the Observer role. All targets (labels, teams, etc. ) return 0 hosts when instead they should return the correct counts for all targets.

This may be because we're currently sending the `queryId` parameter in the `POST /targets` request as a string instead of an integer.

Linking to a Loom video that walks through the above here: https://www.loom.com/share/4303c8a3d21d4b1ab92565b75eccfdb7

martavis commented:
@noahtalerman @zwass All issues reported so far have been resolved.

noahtalerman commented:
@martavis awesome. Thank you for the many hot fixes following feedback

noahtalerman commented:
After pulling down the latest changes, when I select the "Run" button as an observer I see a blank white screen due to a 403 response to the `POST /api/v1/fleet/queries/run` endpoint:
https://www.loom.com/share/a0193825f641491b8818c6d0945f6022

martavis commented:
@noahtalerman I added code to handle Run Query errors, and fixed the hover state for the query's name and description (should not show for observers). My question is do you know why a 403 would show for an observer if the query has `observer_can_run` as true?

noahtalerman commented:
> I added code to gracefully fail on Run Query errors, and fixed the hover state for the query's name and description (should not show for observers)

Boom!

> My question is do you know why a 403 would show for an observer if the query has observer_can_run as true?

@martavis I don't immediately. I think a similar issue was addressed in this PR linked here: https://github.com/fleetdm/fleet/pull/1603

I'll check quickly if a 403 is showing for an observer on the `main` branch.

noahtalerman commented:
> do you know why a 403 would show for an observer if the query has observer_can_run as true?

@martavis ah, I think I have a lead. It looks like we're currently sending the `query` parameter in the `POST /api/v1/fleet/queries/run` for saved queries. We need to send along the `queryId` in order for the backend to determine `observer_can_run` permissions:
https://www.loom.com/share/94442b240ab64daeb17c06891f2fe035

martavis commented:
> I think we need to send the queryId in order for the backend to determine observer_can_run permissions:

Bingo! Thank you for looking into this. I was wondering if it was by design. I think it would be helpful to tweak [the API docs](https://github.com/fleetdm/fleet/blob/main/docs/1-Using-Fleet/3-REST-API.md#run-live-query) a bit to say "Required if running query as an observer" for the `query_id` parameter. I can do it if you agree @noahtalerman.

martavis commented:
@gillespi314 @noahtalerman @zwass @chiiph All checks are passing. I wanted to make sure everyone was good with me merging into the `fleetdm:1497-improved-query-experience` branch for one last PR into `main`.

noahtalerman commented:
> I think it would be helpful to tweak the API docs a bit to say "Required if running query as an observer" for the query_id parameter.

@martavis this makes to me. Please feel free to update the API docs whenever you think clarification is needed :)
